crates/sui-bridge/src/sui_syncer.rs

```diff
@@ -1,14 +1,238 @@
+    cursors: SuiTargetModules,
+}
+
+impl<C> SuiSyncer<C>
+where
+    C: SuiClientInner + 'static,
+{
+    pub fn new(sui_client: Arc<SuiClient<C>>, cursors: SuiTargetModules) -> Self {
+        Self {
+            sui_client,
+            cursors,
+        }
+    }
+
+    pub async fn run(
+        self,
+        query_interval: Duration,
+    ) -> BridgeResult<(
+        Vec<JoinHandle<()>>,
+        mysten_metrics::metered_channel::Receiver<Vec<SuiEvent>>,
+    )> {
+        let (events_tx, events_rx) = mysten_metrics::metered_channel::channel(
+            SUI_EVENTS_CHANNEL_SIZE,
+            &mysten_metrics::get_metrics()
+                .unwrap()
+                .channels
+                .with_label_values(&["sui_events_queue"]),
+        );
+
+        let mut task_handles = vec![];
+        for (module, cursor) in self.cursors {
+            let events_rx_clone = events_tx.clone();
+            let sui_client_clone = self.sui_client.clone();
+            task_handles.push(spawn_logged_monitored_task!(
+                Self::run_event_listening_task(
+                    module,
+                    cursor,
+                    events_rx_clone,
+                    sui_client_clone,
+                    query_interval
+                )
+            ));
+        }
+        Ok((task_handles, events_rx))
+    }
+
+    async fn run_event_listening_task(
+        // The module where interested events are defined.
+        // Moudle is always of bridge package 0x9.
+        module: Identifier,
+        mut next_cursor: TransactionDigest,
+        events_sender: mysten_metrics::metered_channel::Sender<Vec<SuiEvent>>,
+        sui_client: Arc<SuiClient<C>>,
+        query_interval: Duration,
+    ) {
+        tracing::info!(
+            ?module,
+            ?next_cursor,
+            "Starting sui events listening task from tx_digest {next_cursor}"
+        );
+        let mut interval = time::interval(query_interval);
+        interval.set_missed_tick_behavior(time::MissedTickBehavior::Skip);
+        loop {
+            interval.tick().await;
+            // TODO reconsider panic, should we just log an error and continue the loop?
+            let events = retry_with_max_delay!(
+                sui_client.query_events_by_module(PACKAGE_ID, module.clone(), next_cursor),
+                Duration::from_secs(600)
+            )
+            .expect("Failed to query events from sui client after retry");
+
+            let len = events.data.len();
+            if len != 0 {
+                events_sender
+                    .send(events.data)
+                    .await
+                    .expect("All Sui event channel receivers are closed");
+                // Unwrap: `query_events_by_module` always returns Some `next_cursor`
+                // If the events list is empty, `next_cursor` will be the same as `start_tx_digest`
+                next_cursor = events.next_cursor.unwrap();
```

Comment:
Still a bit fragile to depend on a fullnode behavior. We should not unwrap if it's trivial not to.